File: patterns/02_observer/ex2_sensor.go
package main

import (
	"fmt"
	"time"
)

// Step 1: Consider alarm listener is Observer
type AlarmListener interface {
	alarm()
}

type alarmListener struct{}

func (alarmListener) alarm() {
	fmt.Println("Alarm ring")
}

// Step 2: Consider sensor system is Subject
type SensorSystem interface {
	register(sensor AlarmListener)
	remove(sensor AlarmListener)
	soundTheAlarm()
}

// This class refer at:
// https://docs.microsoft.com/en-us/previous-versions/windows/desktop/ipmiprv/sensor
type Sensor struct {
	Caption      string
	Description  string
	ElementName  string
	InstallDate  time.Time
	Name         string
	SensorType   uint16
	CurrentState string
}

type sensorSystem struct {
	alarmListeners []AlarmListener
}

func (s *sensorSystem) register(sensor AlarmListener) {
	s.alarmListeners = append(s.alarmListeners, sensor)
}

func (s *sensorSystem) remove(sensor AlarmListener) {

}

func (s sensorSystem) soundTheAlarm() {

	for i := 0; i < len(s.alarmListeners); i++ {
		sensor := s.alarmListeners[i]
		sensor.alarm()
	}
}

// func RemoveIndex(s []alarmListener, index int) []string {
// 	return append(s[:index], s[index+1:]...)
// }

type Lighting struct{}

// To inheritant methods from AlarmListener
func (Lighting) alarm() {
	fmt.Println("lights up")
}

func NewLighting() AlarmListener {
	return &Lighting{}
}

type Gates struct{}

func (Gates) alarm() {
	fmt.Println("gates close")
}

func NewGates() AlarmListener {
	return &Gates{}
}

type checkList struct{}
type CheckList interface {
	byTheNumbers()
	localize()
	isolate()
	identify()
}

func (c checkList) byTheNumbers() {
	c.localize()
	c.isolate()
	c.identify()
}

func (checkList) localize() {
	fmt.Println("   establish a perimeter")
}

func (checkList) isolate() {
	fmt.Println("   isolate the grid")

}

func (checkList) identify() {
	fmt.Println("   identify the source")
}

type Surveillance struct {
	check checkList
}

func (s Surveillance) alarm() {
	fmt.Println("Surveillance - by the numbers:")
	s.check.byTheNumbers()
}

func NewSurveillance() AlarmListener {
	return &Surveillance{}
}

func main() {
	sensorSystem := &sensorSystem{}
	lighting := NewLighting()
	gates := NewGates()
	surveillance := NewSurveillance()

	sensorSystem.register(lighting)
	sensorSystem.register(gates)
	sensorSystem.register(surveillance)
	sensorSystem.soundTheAlarm()
}

File: patterns/02_observer/ex2_sensor.java
package main;

import java.util.Enumeration;
import java.util.Vector;

//Step 1: Create a interface Alarm Listener
interface AlarmListener {
    void alarm();
}

//Step 2: Implement a Sensor sytem, where register senor
class SensorSystem {
    // Step 2.1: Create a vector mapping, we store all sensor
    private Vector listeners = new Vector();

    public void register(AlarmListener alarmListener) {
        listeners.addElement(alarmListener);
    }

    public void soundTheAlarm() {
        for (Enumeration e = listeners.elements(); e.hasMoreElements();) {
            ((AlarmListener) e.nextElement()).alarm();
        }
    }
}

//Step 3: Implement interface for functions 
class Lighting implements AlarmListener {
    public void alarm() {
        System.out.println("lights up");
    }
}

class Gates implements AlarmListener {
    public void alarm() {
        System.out.println("gates close");
    }
}

//Step 3: Create a class contain manys functions/feature operate on sensor
class CheckList {
 // Template Method design pattern
    public void byTheNumbers() {
        localize();
        isolate();
        identify();
    }

    protected void localize() {
        System.out.println("   establish a perimeter");
    }

    protected void isolate() {
        System.out.println("   isolate the grid");
    }

    protected void identify() {
        System.out.println("   identify the source");
    }
}

//class inherit.
//type inheritance
class Surveillance extends CheckList implements AlarmListener {
    public void alarm() {
        System.out.println("Surveillance - by the numbers:");
        byTheNumbers();
    }

    protected void isolate() {
        System.out.println("   train the cameras");
    }
}

public class ex2_sensor {
 public static void main( String[] args ) {
     SensorSystem sensorSystem = new SensorSystem();
     sensorSystem.register(new Gates());
     sensorSystem.register(new Lighting());
     sensorSystem.register(new Surveillance());
     sensorSystem.soundTheAlarm();
 }
}
